src/librustc_typeck/astconv.rs

```diff
@@ -2305,13 +2313,36 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
             item_def_id,
             item_segment,
             trait_ref.substs,
+            generic_arg_position,
         );
 
         debug!("qpath_to_ty: trait_ref={:?}", trait_ref);
 
         self.normalize_ty(span, tcx.mk_projection(item_def_id, item_substs))
     }
 
+    pub fn prohibit_multiple_params<'a, T: IntoIterator<Item = &'a hir::PathSegment<'a>>>(
+        &self,
+        segments: T,
+    ) -> bool {
+        let segments_with_params = segments
+            .into_iter()
+            .filter_map(|segment| segment.args.map(|arg| arg.span))
+            .collect::<Vec<_>>();
+        if segments_with_params.len() <= 1 {
+            return false;
+        }
+        self.tcx()
+            .sess
+            .struct_span_err(
+                segments_with_params,
+                "multiple segments with type parameters are not allowed",
+            )
+            .note("only a single path segment can specify type parameters")
+            .emit();
+        true
+    }
```

Comment:
I think this is unnecessarily complicated, it doesn't distinguish this "`enum` vs variant" situation from others (e.g. trying to pass generic args to a `mod`) and it's wrong with generic associated items like methods or GATs.

This could be a help message on the error from `prohibit_generics`, such as pointing to the `enum` segment and saying "this segment *already* specified generic args for the `enum`" or something.

Reply:
> I think this is unnecessarily complicated, it doesn't distinguish this "enum vs variant" situation from others (e.g. trying to pass generic args to a mod) and it's wrong with generic associated items like methods or GATs.

That check [is being done when called](https://github.com/rust-lang/rust/pull/69363/files#diff-1d1b0d29a2e8da97c6bfb6e364d920c7R5420). If it were called without that check it would indeed be incorrect.

> This could be a help message on the error from prohibit_generics, such as pointing to the enum segment and saying "this segment already specified generic args for the enum" or something.

Wanted to separate this from E0109 to potentially give it its own error code to explain that this will never be supported as opposed to the more generic current error.

Reply:
> to explain that this will never be supported

What do you mean? Generics on both the `enum` and a variant is a plausible feature (GADT), and generics on modules have been talked about in the past.

The error message looks pretty misleading to me, it seems to claim some absolute truth.